pyext/fastfunc.c

```diff
@@ -89,7 +89,10 @@ func_Utf8DecodeOne(PyObject *self, PyObject *args) {
   if (!PyArg_ParseTuple(args, "s#i", &string, &length, &start)) {
     return NULL;
   }
-  // Bounds check for safety
+
+  // Bounds check for safety. start can equal length because the string has a
+  // nul-terminator. So utf8_decode(s + len(s)) is valid and would decode that
```

Comment:
Hm actually the fact that you would get a zero code point from decoding the terminator seems bad !!!  That NUL terminator is not part of the string, just like it's not in C.

Or maybe it's OK because the caller should always check the error and not use the code point zero ?

---

It still seems a bit confusing but I don't want to have us blocked on this, since it's probably not causing a problem now


Maybe I need to look at the tests more closely

I think it's possible the unit tests may be asserting some properties that we don't need in real code, i.e. the `s[len(s)]` case

Reply:
Yeah, there's some subtlety around zero-codepoints. I *cannot* assume inside `utf8_decode` that those denote the end of a string. Instead, we return both that the decoded codepoint is zero and that this *may* be the end of the string. But, like the surrogate errors, the caller can decide if that's fatal or not. We need this because, unlike `data_lang/j8.h`, `osh/string_ops.py` is perfectly fine with zero-codepoints in the middle of the string.

Although, I did add the `mylib.ByteAt(s, i) == 0` check in `osh/string_ops.py`... so `utf8_decode` doesn't actually see any nul-terminators in the middle of the code. I should probably do this instead:

```python
codepoint_or_error, _bytes_read = fastfunc.Utf8DecodeOne(s, i)
if codepoint_or_error == UTF8_ERR_END_OF_STREAM:
  # utf8_decode found a nul-terminator, but we allow those in the middle of our strings
  return 0
elif codepoint_or_error < 0:
  raise error
return codepoint_or_error
```

Reply:
Oh hm I didn't look carefully at the special case for NUL.

I think if we can get rid of that we should ... The definition of an Oils/YSH string is the same as a Python 2 string basically -- it's an array  of bytes, which can include NULs, which may or may not be UTF-8 encoded

That is for PURE data operations like string trim() etc.

BUT when you make a syscall like `chdir()`, then Unix does NUL terminated strings, so it's OK to truncate in that case. 

But for regular string processing we shouldn't treat it as special.

----

j8.h seems to deal with NUL fine, without any special cases

```
ysh andy@andy-VirtualBox:~/git/oilshell/oil$ var x = b'a\y00\y00b'

ysh andy@andy-VirtualBox:~/git/oilshell/oil$ = x
(Str)   "a\u0000\u0000b"

ysh andy@andy-VirtualBox:~/git/oilshell/oil$ json write (x)
"a\u0000\u0000b"
```

The subtlety is that

- we rely on the trailing NUL only for INVALID sequences.  If you have a 4 byte sequence that's cut off at 3 bytes, we don't know until we hit the NUL.  We need teh NUL.
- we do NOT rely on it for VALID sequences.  There we can rely on the length check

So yeah I wrote that big comment in `j8.h` because I found this confusing. 

---

But I think we should try to design the API to get rid of that special case in Python -- it will affect the forward string iterator too

---

I also think if we never actually access `s[len(s)]`, and can rely on the length check, we should not handle that case and make the assertion fail

That is I think we should distinguish our Python-level APIs from the C one

- utf8decode() is  a low level C helper.
- J8 encoding, utf-8 validity, code point iteration are Python-level APIs

I think it's more important to be strict about the Python APIs, which use GC strings, with internal NULs

And utf8decode() is basically an implementation detail, which uses C strings which are buffers.